src/main/java/com/team1806/frc2020/subsystems/Turret.java

```diff
@@ -0,0 +1,152 @@
+package com.team1806.frc2020.subsystems;
+
+import com.ctre.phoenix.motorcontrol.ControlMode;
+import com.ctre.phoenix.motorcontrol.can.TalonSRX;
+import com.team1806.frc2020.Constants;
+import com.team1806.lib.util.ReflectingCSVWriter;
+import edu.wpi.first.wpilibj.Timer;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+
+
+public class Turret extends Subsystem {
+
+    enum TurretControlState {
+
+        kIdle, kPositionControl, kHoldPosition, kManualControl;
+    }
+
+
+    private class PeriodicIO {
+        public double timestamp;
+        public double motorVoltage;
+        public double motorAmps;
+        public double currentPosition;
+        public double currentSpeed;
+        public double wantedPosition;
+        public double wantedManualMovement;
+        public TurretControlState turretState;
+
+    }
+
+
+
+    private static Turret TURRET = new Turret();
+
+    private TurretControlState mTurretControlState;
+    private TalonSRX mCANTalonSRX;
+    private PeriodicIO mPeriodicIO;
+    private ReflectingCSVWriter mCSVWriter;
+
+
+    private Turret(){
+        mCANTalonSRX = new TalonSRX(Constants.kTurretMotor);
+        mPeriodicIO = new PeriodicIO();
+    }
+
+    public static Turret GetInstance(){
+        return TURRET;
+    }
+
+    private double ConvertEncoderClicksToAngle(double encoderCount){
+        return encoderCount * Constants.kTurretDegreesPerCount;
+    }
+
+    private double ConvertAngleToEncoderClicks(double angle){
+        return angle / Constants.kTurretDegreesPerCount;
+    }
+
+    public void setWantedAngle(double angle){
+        setControlState(TurretControlState.kPositionControl);
+        mPeriodicIO.wantedPosition = angle;
+    }
+
+    public void setManualControl(double wantedPosition){
+        setControlState(TurretControlState.kManualControl);
+        mPeriodicIO.wantedManualMovement = wantedPosition;
+    }
+
+    public void setWantIdle(){
+        setControlState(TurretControlState.kIdle);
+        mPeriodicIO.wantedPosition = 0.0;
+        mPeriodicIO.wantedManualMovement = 0.0;
+    }
+
+    public boolean isOnTarget() {
+        return Math.abs(mPeriodicIO.currentPosition) >= Math.abs(mPeriodicIO.wantedPosition - Constants.kTurretAcceptableAngleDeviation) && mPeriodicIO.currentSpeed <= Constants.kTurretAcceptableSpeed;
+    }
+
+
+
+    public void writePeriodicOutputs(){
+        switch (mPeriodicIO.turretState){
+            default:
+            case kIdle:
+                mCANTalonSRX.set(ControlMode.Position, 0);
+                break;
+            case kPositionControl:
+                mCANTalonSRX.set(ControlMode.Position, mPeriodicIO.wantedPosition);
+                break;
+            case kHoldPosition:

```

Comment:
Hold position needs to do something, I suggest a dumb proprtional loop basically just stalling the motor, see our 2018/2019 lift code.